### PR TITLE
simplify capture of clone vars

### PR DESCRIPTION
**Pull request by sitano**
Instead of doing

```
let x = Rc<...>(...);
let f = || x.clone();
let f2 = f.clone()
```

It can be done with anonymous function like this:

```
let second = local_ex
            .spawn_into(
                (|f: Rc<RefCell<bool>>, s: Rc<RefCell<bool>>| {
                    async move {
                       ...
                    }
                })(first_started.clone(), second_status.clone()),
                tq2,
            )
            .unwrap();

```

Rust can't derive type information for (f, s) himself
so it's required to be stated it explicitly as in
`|f: Rc<RefCell<bool>>, s: Rc<RefCell<bool>>|` instead of `|f, s|`.

However, there is a popular macro `enclose!` that does the job.

```
  enclose! { (x, y)
    async move {
    ...
    }
  }
```

### CODE REVIEW

**Comment by sitano** [comment]
@glommer I am sorry GitHub does not allow reopening force pushed commits.

**Review by glommer — CHANGES_REQUESTED** [review]

File: scipio/src/executor.rs [hunk 1]
```diff
@@ -1093,10 +1096,8 @@ fn task_with_latency_requirements() {
 
 #[test]
 fn current_task_queue_matches() {
-    use crate::Local;
-    use futures::join;
-
     let local_ex = LocalExecutor::new(None).unwrap();
+
```
[the remark this answers is not in the document]
Comment (sitano) [reply scipio/src/executor.rs:1100]: Ok, but all the rest of the code has a new line separating local_ex val and a call to run

File: scipio/src/executor.rs [hunk 2]
```diff
@@ -45,9 +45,15 @@ use scoped_tls::scoped_thread_local;
 use crate::multitask;
 use crate::parking;
 use crate::task::{self, waker_fn::waker_fn};
+use crate::Local;
 use crate::Reactor;
 use crate::{IoRequirements, Latency};
 
+#[cfg(test)]
```
[the remark this answers is not in the document]
Comment (sitano) [reply scipio/src/executor.rs]: exported lib.rs: `pub use enclose::enclose;`
use in tests: `crate::enclose!`

File: scipio/src/executor.rs [hunk 3]
```diff
@@ -45,9 +45,15 @@ use scoped_tls::scoped_thread_local;
...
+use enclose::enclose;
+#[cfg(test)]
+use futures::join;
```
[the remark this answers is not in the document]
Comment (sitano) [reply scipio/src/executor.rs]: removed. As there is no sense to do a single usage `use` a left `futures::join!`

**Review by glommer — COMMENTED** [review]
Hi Ivan

Good progress, we are getting ready to merge status here!

As I had said previously, I would like to see a very simple example of some external code using this to make sure we don't break it. Exporting macros is tricky.

something under `examples` that uses this.

The other comment is that this patch is mixing adding enclose with lots of cleanups in which you are changing Task:: to Local.

Those cleanups are very welcome but I would like to see them split at a separate patch. You can leave that for the end if you prefer to only add the example in the next round

**Comment by sitano** [comment]
@glommer I have added ping pong example:

```
$ cargo build --example ping_pong
$ ./target/debug/examples/ping_pong

right
left
reset
...
```

**Review by glommer — APPROVED** [review]
Ivan, I am approving 960bce3dee3c17268056d8fd3eca32c9921f2a3e

but I would like to see this split in two patches: the first one only with the cleanups like changing task to local, and then on top you can add your enclose.

Please force push a version that does that, and make sure that the diff from 960bce3dee3c17268056d8fd3eca32c9921f2a3e comes empty.

**Comment by sitano** [comment]
ok!

**Comment by sitano** [comment]
@glommer split into:

- cleanup: use Local instead of Task::<()>
- clone captured vars for async closures …

**Pull request merged**